### ISSUE TITLE
LB rule with selector removed when any labelled service removed

### ISSUE THREAD

**Issue by VAdamec** [rancher/rancher#7892]
**Rancher Versions:**
Server: 1.4.1
 - Rancher/lb-service-haproxy:v0.5.9

**Docker Version:**
1.12.6

**OS and where are the hosts located? (cloud, bare metal, etc):**
openstack, centos7

**Setup Details: (single node rancher vs. HA rancher, internal DB vs. external DB)**
Rancher HA with external DB (galera)

**Environment Type: (Cattle/Kubernetes/Swarm/Mesos)**
Cattle

**Steps to Reproduce:**
 - Balancer is used like global service (not deployed for specific stack/service)
 - LB service -> Add Selector Rule -> https,443 and target: loadbalancer.https=true
start two services with label 'loadbalancer.https: true' add lb_config

```
    lb_config:
      port_rules:
      - hostname: first.example.com
        target_port: 8080
```

```
    lb_config:
      port_rules:
      - hostname: second.example.com
        target_port: 8080
```

remove one of services, check rules on LB.

**Results:** 
Selector rule is removed with service, expected is that this rule will stay here (at least until all services with this label not removed).

**Comment by tcdev0** [comment]
+1 same here

**Comment by mhdev90** [comment]
+1 same issue.

Rancher Version:
* Server: 1.4.2
* rancher/lb-service-haproxy:v0.5.9

Docker Version:
* 1.12.6

**Comment by VAdamec** [comment]
Same for 1.5.1 and rancher/lb-service-haproxy:v0.6.2, Docker v1.13.1

**Comment by deniseschannon** [comment]
@VAdamec It would help me if you could put your full compose files for the load balancers and services to it'd be easier to reproduce and ensure we're doing the same steps. 

Could anyone give more details on how you are adding/removing services? Was it through the UI or compose? How are you checking the rules?

Are there additional rules or just 1 selector rule on the LB?

I assume your 2 services were started using CLI/rancher-compose as there is no UI support for adding in port rules to services being targeted by selectors.

I removed the service through the UI. 

I am unable to reproduce in v1.5.3 using `lb-service-haproxy:v0.6.2`

Steps I did:

1. I started a load balancer using selector rules with 2 services being picked up by doing "Add Stack and inputting the following compose"

docker-compose.yml

``` 
version: '2'
services:
  lb1:
    image: rancher/lb-service-haproxy:<version>
    ports:
    - 81
  web1:
    image: nginx
    labels:
      foo: bar
  web2:
    image: nginx
    labels:
        foo: bar
```

rancher-compose.yml
```
version: '2'
services:
  lb1:
    scale: 1
    lb_config:
      port_rules:
      - source_port: 81
        # Target any service that has foo=bar as a label
        selector: foo=bar
        protocol: http
    health_check:
      port: 42
      interval: 2000
      unhealthy_threshold: 3
      healthy_threshold: 2
      response_timeout: 2000
  # web1 and web2 are targeted with the same source port but with the different hostname and path rules
  web1:
    scale: 1
    lb_config:
      port_rules:
      - target_port: 80
        hostname: test.com
  web2:
    scale: 1
    lb_config:
      port_rules:
      - target_port: 80
        hostname: example.com/test     
```

2. From the UI, I can see the load balancer and 2 services. I check the yaml of the stack by clicking on "View Config" from the dropdown of the stack. I also check the LB that the HAProxy is programmed correctly for both services by execing into the LB container and viewing `/etc/haproxy/haproxy.cfg`

3. From the UI, I delete one of the services. I re-check the yaml of the stack and the

```
selector: foo=bar
``` 

part of the lb_config is still in.

I check the HAProxy and confirm that there is still 1 service being targetted.

I tried these steps with a load balancer with a specific scale as well a load balacner with a global scale. Both times, after deleting one of 2 services, the port rules on the load balancer still remained.

**Comment by VAdamec** [comment]
I'll attach full files, but there is nothing special. Selector rule was added manually via UI, all stack/service changes are done via rancher-compose and checked in UI (or test by app test suite and rancher-cli). But I have same result If I remove stack/service via UI.

**Comment by deniseschannon** [comment]
Where or how are you checking that the selector rule still exists on the LB?

I just tried deleting the stack in the UI and my selector rule is still on the LB.

**Comment by VAdamec** [comment]
Also good to mention, LB are not part of app stack, it's separate service. So If you need to test it you need to separate LB and web1/2 services to docker/rancher config files. 

Idea behind this usage is quite simple, in corporate company you cannot just simple open all ports on all servers/ports to all clients, so you select special separate range just for LB services, which has special firewall rules. Other hosts are not used for publishing external ports. So all 80/443 communication are setup in one stack/LB service - one selector rule for all. Same approach is used for example with Traefic.

**Comment by deniseschannon** [comment]
Yes, I had tried it where the LB is not part of the app stack and it's a separate service and tried deleting the app separately. I am still able to see the LB selector rule on my LB.

**Comment by VAdamec** [comment]
LB compose files: https://gist.github.com/VAdamec/315c57273626fede1a1ce19a36b5e054

**Comment by VAdamec** [comment]
I made completely fresh environment, add LB config and add service, with this setup https://gist.github.com/VAdamec/f5a89dd36b76d535bad69abe6a12df0a

and yes, rule stay even if I delete last service which use it. So question is what's different. I'll remove and create LB in production environment to see If this helps or not.

**Comment by VAdamec** [comment]
Ok so that helps, I removed LB config and create exactly same configuration and no issue with selector rule now. Maybe some issue during upgrading between Rancher version?

**Comment by mattdodge** [comment]
I'm going to request this be re-opened. I'm experiencing the same thing (and it seems others at the top of the thread are too) and I don't really consider "just start over" to be a sufficient resolution to the bug that likely exists.

Here are my semi-anonymized docker-compose/rancher-compose files for the load balancer that is problematic: https://gist.github.com/mattdodge/a0755d3b42538314449b4183d7993d33

Like @VAdamec, this stack is separate from the stacks that the load balancer points to. So the bug that is happening is if I were to delete the service `855f7312-ad55-4908-a8bc-6e5f3245f519/image-1` then the rule for that service **and** the selector rule (`rule.proxyable=instance`) would both be deleted. The other service rule remains. The desired behavior would be for both the other service and the selector rule to exist on the load balancer.

If this doesn't get reopened I can create another issue. If I do that, I will link it to this one so at least there is some trail.

**Comment by deniseschannon** [comment]
How are you launching services/deleting them? As I mentioned earlier I was unable to reproduce but I need clearer steps so I can follow exactly what you guys are doing. 

For launching, are you using UI, compose/CLI or catalog?
Are you launching the LB or services first?
For deleting, are you using the UI or CLI?

**Comment by mattdodge** [comment]
Launching services via an API call (POST to `/stacks` endpoint). Deleting them via the UI for now. I think in all cases the load balancer was set up before the services, then the rules were added to the load balancer once the services were started.

**Comment by VAdamec** [comment]
After upgrade to 1.6.0 I'm get this issue back (rancher/lb-service-haproxy:v0.7.1)

**Comment by carminexx** [comment]
Same here; Rancher v1.6.2

**Comment by occasl** [comment]
We're experiencing the issue as well in Rancher v1.6.x.

**Comment by stszap** [comment]
@deniseschannon We are seeing the same problem, it's easy to reproduce. I tested with rancher 1.6.4 on local machine and tried to make it as simple as possible:

1. created "Default" stack from ui
2. created "nginx" service from ui (name "nginx", image "nginx")
3. created load balancer "lb" from ui with 1 rule (port 9000, target "nginx", target port 80)
4. checked load balancer configuration via "Upgrade/Edit" button in the ui (my rule is there)
5. deleted "nginx" service
6. checked load balancer configuration again (no rules)
7. created "nginx" service again 
8. checked load balancer configuration again (still no rules)

**Comment by vincent99** [comment]
@stszap this issue is talking about selector rules, which indirectly target containers that have a label.

You're taking about rules directly pointed at a service, which are correctly (intentionally) removed if the service is removed instead of leaving garbage references pointing to nonexistent services lying around.  They are linked by ID anyway, not name, so adding a new service with the same name wouldn't be associated with the old rule even if it still existed.

**Comment by stszap** [comment]
@vincent99 Oh, thank you for clarification.

**Comment by aemneina** [comment]
I believe this is related to #9118 and fixed in 1.6.6
